File: haskell/Scene.hs
module Scene (Scene(..), fromScene,
              rgbbox, irreg) where

import BVH
import Raytracing
import Vec3

data Scene = Scene { sceneCamLookFrom :: Pos
                   , sceneCamLookAt :: Pos
                   , sceneCamFov :: Float
                   , sceneSpheres :: [Sphere]
                   }
           deriving (Show, Read)

fromScene :: Int -> Int -> Scene -> (Objs, Camera)
fromScene width height scene =
  (mkBVH sphereAABB $ sceneSpheres scene,
   mkCamera (sceneCamLookFrom scene) (sceneCamLookAt scene) (Vec3 0 1 0)
   (sceneCamFov scene) (fromIntegral width/fromIntegral height))

rgbbox :: Scene
rgbbox = Scene { sceneSpheres = leftwall ++ midwall ++ rightwall ++ bottom
               , sceneCamLookFrom = Vec3 0 30 30
               , sceneCamLookAt = Vec3 0 (-1) (-1)
               , sceneCamFov = 75 }
  where n = 10
        k = 60
        leftwall =
          [ Sphere (Vec3
                    (-k/2)
                    (-k/2 + (k/n) * y)
                    (-k/2 + (k/n) * z))
            (Vec3 1 0 0) (k/(n*2))
          | y <- [0..n-1], z <- [0..n-1]
          ]
        midwall =
          [ Sphere (Vec3
                     (-k/2 + (k/n) * x)
                     (-k/2 + (k/n) * y)
                     (-k/2))
            (Vec3 0 1 0) (k/(n*2))
          | x <- [0..n-1], y <- [0..n-1]
          ]
        rightwall =
          [ Sphere (Vec3
                     (k/2)
                     (-k/2 + (k/n) * y)
                     (-k/2 + (k/n) * z))
            (Vec3 0 0 1) (k/(n*2))
          | y <- [0..n-1], z <- [0..n-1]
          ]
        bottom =
          [ Sphere (Vec3
                     (-k/2 + (k/n) * x)
                     (-k/2)
                     (-k/2 + (k/n) * z))
            (Vec3 1 1 1) (k/(n*2))
          | x <- [0..n-1], z <- [0..n-1]
          ]

irreg :: Scene
irreg = Scene { sceneSpheres = bottom
              , sceneCamLookFrom = Vec3 0 12 30
              , sceneCamLookAt = Vec3 0 10 (-1)
              , sceneCamFov = 75 }
  where n = 100
        k = 600
        bottom =
          [ Sphere (Vec3
                     (-k/2 + (k/n) * x)
                     0
                     (-k/2 + (k/n) * z))
            (Vec3 1 1 1) (k/(n*2))
          | x <- [0..n-1], z <- [0..n-1]
          ]

File: scala/src/main/scala/raytracer/Scene.scala
package raytracer

import Raytracer.{Pos, Sphere, Camera, Objs}

final case class Scene(camLookFrom: Pos, camLookAt: Pos, camFov: Double, spheres: List[Sphere]) {
  def toObjsCam(width: Int, height: Int): (Objs, Camera) =
    (BVH(_.aabb, spheres), Camera(camLookFrom, camLookAt, Vec3(0,1,0), camFov, width.toDouble / height.toDouble))
}

object Scene {

  def fromString(name: String): Option[Scene] = name match {
    case "irreg" => Some(irreg)
    case "rgbbox" => Some(rgbbox)
    case _ => None
  }

  def rgbbox: Scene = {
    val n = 10
    val k = 60.0
    val leftwall = for {
      y <- 0 to n-1
      z <- 0 to n-1
    } yield 
    Sphere( Vec3( -k/2.0
                , -k/2.0 + k/n * y
                , -k/2.0 + k/n * z)
          , Vec3(1,0,0)
          , k/(n * 2.0)
          )

    val midwall = for {
      x <- 0 to n-1
      y <- 0 to n-1
    } yield 
    Sphere( Vec3( -k/2.0 + k/n * x
                , -k/2.0 + k/n * y
                , -k/2.0)
          , Vec3(0,1,0)
          , k/(n * 2.0)
          )
    val rightwall = for {
      y <- 0 to n-1
      z <- 0 to n-1
    } yield 
    Sphere( Vec3(  k/2.0
                , -k/2.0 + k/n * y
                , -k/2.0 + k/n * z)
          , Vec3(0,0,1)
          , k/(n * 2.0)
          )

    val bottom = for {
      x <- 0 to n-1
      z <- 0 to n-1
    } yield
    Sphere( Vec3( -k/2.0 + k/n * x
                , -k/2.0
                , -k/2.0 + k/n * z)
          , Vec3(1,1,1)
          , k/(n*2.0)
          )

    Scene( Vec3(0, 30, 30)
         , Vec3(0, -1, -1)
         , 75
         , (leftwall ++ midwall ++ rightwall ++ bottom).toList
         )
  }

  def irreg: Scene = {
    val n = 100
    val k = 600.0
    val bottom = for {
      x <- 0 to n-1
      z <- 0 to n-1
    } yield
    Sphere( Vec3( -k/2.0 + k/n * x
                , 0
                , -k/2.0 + k/n * z)
          , Vec3(1,1,1)
          , k/(n*2.0)
          )

    Scene(Vec3(0, 12, 30), Vec3(0, 10, -1), 75, bottom.toList)
  }
}

